### PR TITLE
Improve nginx-ingress helm chart upgrade process.

### LINKED ISSUE #8700: nginx-ingress upgrade does not revert to previous state on failure
### What happened?

When upgrading from KKP2.18.4 to KKP2.19 using the `--migrate-upstream-nginx-ingress` flag, @dermorz and I noticed that when a helm release in the ingress-controller fails, its deployment is deleted and it will not be restored. This in turn blocks all future installs for the kkp-installer

### What should happen?

Install succeeds or at least the deleted deployment should be recovered from backup

### How to reproduce?

Run the installer, but have an invalid nginx values.yaml section (see example below)

```yaml
# example for broken section (podLabels should be keyvalue instead of list)
nginx:
    controller:
        podLabels:
            - "kkptest": itworks
            - "anotherone": yesyes
```

This will trigger our deployment deletion logic and afterwards expectedly fail when running the release. So far so good.

```
./kubermatic-installer219 deploy --config kubermatic-configuration.yaml --helm-values values.yaml --storageclass aws --migrate-upstream-nginx-ingress --migrate-upstream-cert-manager


INFO[10:07:12] 🛫 Initializing installer…                     edition="Enterprise Edition" version=v2.19.0-alpha.4
INFO[10:07:12] 🚦 Validating the provided configuration…
WARN[10:07:12]    Helm values: kubermaticOperator.imagePullSecret is empty, setting to spec.imagePullSecret from KubermaticConfiguration
INFO[10:07:12] ✅ Provided configuration is valid.
I0112 10:07:14.039016   34235 request.go:665] Waited for 1.040396968s due to client-side throttling, not priority and fairness, request: GET:https://pgdvnvrfnd.kubermatic.test-env-kkp.ee.lab.kubermatic.io:31305/apis/crd.projectcalico.org/v1?timeout=32s
INFO[10:07:14] 🧩 Deploying KKP master stack…
INFO[10:07:14]    💾 Deploying kubermatic-fast StorageClass…
INFO[10:07:14]    ✅ StorageClass exists, nothing to do.
INFO[10:07:14]    📦 Deploying nginx-ingress-controller…
INFO[10:07:14]       nginx-ingress-controller: 1.2.2 detected, performing upgrade to 1.3.4…
INFO[10:07:14]       Backing up old ingress deployment…
INFO[10:07:14]       attempting to store the deployment
INFO[10:07:14]       Deleting the deployment from the cluster
INFO[10:07:14]       Deploying Helm chart…
INFO[10:07:14]       Updating release from 1.2.2 to 1.3.4…
ERRO[10:07:18] ❌ Operation failed: failed to deploy nginx-ingress-controller: failed to deploy Helm release: failed to install: coalesce.go:199: warning: cannot overwrite table with non table for podLabels (map[])
coalesce.go:199: warning: cannot overwrite table with non table for podLabels (map[])
coalesce.go:199: warning: cannot overwrite table with non table for podLabels (map[])
Error: UPGRADE FAILED: template: nginx-ingress-controller/templates/service.yaml:26:22: executing "nginx-ingress-controller/templates/service.yaml" at <.Values.nginx.service.annotations>: nil pointer evaluating interface {}.annotations.

```

**However if you run it now with the correct values (see below), the installer will fail earlier because the deployment does not exist anymore (it got deleted and was never restored by the previous run)**

```yaml
nginx:
    controller:
        podLabels:
            "kkptest": itworks
            "anotherone": yesyes
```

```
❯ ./kubermatic-installer219 deploy --config kubermatic-configuration.yaml --helm-values values.yaml --storageclass aws --migrate-upstream-nginx-ingress --migrate-upstream-cert-manager

INFO[10:08:27] 🛫 Initializing installer…                     edition="Enterprise Edition" version=v2.19.0-alpha.4
INFO[10:08:27] 🚦 Validating the provided configuration…
WARN[10:08:27]    Helm values: kubermaticOperator.imagePullSecret is empty, setting to spec.imagePullSecret from KubermaticConfiguration
INFO[10:08:27] ✅ Provided configuration is valid.
I0112 10:08:28.360643   34836 request.go:665] Waited for 1.036590594s due to client-side throttling, not priority and fairness, request: GET:https://pgdvnvrfnd.kubermatic.test-env-kkp.ee.lab.kubermatic.io:31305/apis/acme.cert-manager.io/v1?timeout=32s
INFO[10:08:28] 🧩 Deploying KKP master stack…
INFO[10:08:28]    💾 Deploying kubermatic-fast StorageClass…
INFO[10:08:28]    ✅ StorageClass exists, nothing to do.
INFO[10:08:28]    📦 Deploying nginx-ingress-controller…
INFO[10:08:28]       nginx-ingress-controller: 1.2.2 detected, performing upgrade to 1.3.4…
INFO[10:08:28]       Backing up old ingress deployment…
INFO[10:08:28]       attempting to store the deployment
ERRO[10:08:28] ❌ Operation failed: failed to deploy nginx-ingress-controller: failed to upgrade nginx-ingress-controller: found more than one deployment (0) matching the nginx-ingress-controller release, stopping upgrade....
```

From the general installer user-experience we think, that either:
* we should roll-back the deployment on a failed helm release
* (maybe) we can change our strategy to first do the helm release and then just delete the pods to get them re-created
(only meant as possible suggestion, we are not too deep in the ideas that went into the implementation)

### Environment

- KKP Version: 2.18.4 to 2.19.alpha4
- Shared or separate master/seed clusters?: shared
- Master/seed cluster provider: AWS

### PR DESCRIPTION
**Pull request by wurbanski**
**What does this PR do / Why do we need it**:
* Show the name of backup file in the logs
* notify the user to either redeploy previous deployment or fix the errors before proceeding

**Does this PR close any issues?**:

Fixes #8700

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/install/stack/kubermatic-master/nginxingress.go [hunk 1]
```diff
@@ -123,23 +129,26 @@ func upgradeNginxIngress(
 	}); err != nil {
 		logger.Warn("Error querying API for the existing deployment, attempting to upgrade without removing it...")
 	} else {
```
Comment (SimonTheLeg) [inline pkg/install/stack/kubermatic-master/nginxingress.go]: _Unfortunately GH does not let me mark the whole area, but the following comment refers to line 125-131:_

This reminds me of another PR for cert-manager that I reviewed (#8392), and I think the same applies here: We should correctly catch API query issues. Additionally I think we can also simplify the code, by using one less else branch. I had something like this in mind

```go
// can of course also be the "if err..." one liner, no strong preference here
err := kubeClient.List....
if err != nil {
	// I think we always should catch that, as it is like a network error and not a "cannot find deployment"
	return fmt.Errorf("Error querying API for the existing deployment, aborting deployment")
}

switch foundDeployments....
```

**Review by SimonTheLeg — CHANGES_REQUESTED** [review]
Unfortunately I was not able to test this yet, as QA is currently having issues creating new clusters. Will get back on it tomorrow morning.

Apart from that, there was one thing I noticed, that we can change already I think

File: pkg/install/stack/kubermatic-master/nginxingress.go (hunk 1 above)
Comment (wurbanski) [reply pkg/install/stack/kubermatic-master/nginxingress.go]: Definitely makes sense, I will change that as well :)

**Comment by xrstf** [comment]
Didn't we talk about purging the chart? Or am I mixing something up here?

**Comment by wurbanski** [comment]
> Didn't we talk about purging the chart? Or am I mixing something up here?

That was for cert-manager, ingress seemed to be a little more straightforward to upgrade. This PR basically handles the negative path better.

**Review by xrstf — APPROVED** [review]
/approve

**Comment by SimonTheLeg** [comment]
I have verified this PR by doing the following:

1. Setup a quick KKP 2.18.2 installation using the 2.18.2 installer
2. Add a broken nginx section to the `values.yaml` file
```yaml
nginx:
    controller:
        replicaCount: "thisshouldfail"
        service:
            annotations: "thisshouldfail"
```
3. Install using the installer from this branch (with ./charts dir from 2.19 release)
```
./kubermatic-installer-dev deploy --config kubermatic-configuration.yaml --helm-values values_219.yaml --storageclass aws --migrate-upstream-nginx-ingress --migrate-upstream-cert-manager
...

INFO[14:38:38]    📦 Deploying nginx-ingress-controller…
INFO[14:38:38]       nginx-ingress-controller: 1.2.2 detected, performing upgrade to 1.3.4…
INFO[14:38:38]       Backing up old ingress deployment…
INFO[14:38:38]       Attempting to store the deployment in file: backup_nginx-ingress-controller_2022-01-17T143838.yaml
INFO[14:38:38]       Deleting the deployment from the cluster
INFO[14:38:38]       Deploying Helm chart…
INFO[14:38:38]       Updating release from 1.2.2 to 1.3.4…
ERRO[14:38:42] ❌ Operation failed: failed to deploy nginx-ingress-controller: failed to deploy Helm release: failed to install: 
coalesce.go:199: warning: cannot overwrite table with non table for annotations (map[helm.sh/resource-policy:keep])
coalesce.go:199: warning: cannot overwrite table with non table for annotations (map[])
coalesce.go:199: warning: cannot overwrite table with non table for annotations (map[helm.sh/resource-policy:keep])
coalesce.go:199: warning: cannot overwrite table with non table for annotations (map[])
Error: UPGRADE FAILED: template: nginx-ingress-controller/charts/nginx/templates/controller-service.yaml:6:35: executing "nginx-ingress-controller/charts/nginx/templates/controller-service.yaml" at <.Values.controller.service.annotations>: range can't iterate over thisshouldfail

use backup file to restore the deployment or re-try the installation after fixing any errors..
```

Now the error message will be shown. We could slightly make it nicer by printing out the file location in the "use backup file to restore the deployment or re-try the installation after fixing any errors.." log line, but overall this works now as intended 👍

**Review by SimonTheLeg — APPROVED** [review]

**Comment by xrstf** [comment]
/override pre-kubermatic-test-integration

vCenter issues.

**Comment by xrstf** [comment]
/override pre-kubermatic-test-integration

vCenter issues.

**Pull request merged**